src/firefly/java/edu/caltech/ipac/firefly/server/dpanalyze/SofiaAnalyzer.java

```diff
@@ -0,0 +1,285 @@
+            URL url = new URL("http://web.ipac.caltech.edu.s3-us-west-2.amazonaws.com/staff/roby/demo/wise-00.fits");
+            URLDownload.getDataToFile(url, f);
+            FileAnalysisReport retRep = inputReport.copy(false);
+            FileAnalysisReport.Part replacePart =
+                    new FileAnalysisReport.Part(FileAnalysisReport.Type.Image, "A image to replace");
+            replacePart.setFileLocationIndex(0);
+            replacePart.setDefaultPart(true);
+            replacePart.setUiRender(FileAnalysisReport.UIRender.Image);
+            replacePart.setUiEntry(FileAnalysisReport.UIEntry.UseSpecified);
+            replacePart.setConvertedFileName(ServerContext.replaceWithPrefix(f));
+            retRep.addPart(replacePart);
+            return retRep;
+        } catch (MalformedURLException | FailedRequestException e) {
+            e.printStackTrace();
+            return inputReport;
+        }
+    }
+
+    private FileAnalysisReport addingSpectraExtractedTableAPart(FileAnalysisReport inputReport, Sofia1DSpectraExtractor model) {
+        try {
+
+            File spectra = extractSpectraTable(inputReport.getFilePath(), model);
+            FileAnalysisReport retRep = inputReport.copy();
+            DataGroup dg = retRep.getPart(0).getDetails();
+            String spectraName = "Extracted Data ";
+            for (int i = 0; i <dg.size();i++) {
+                String key = (String) dg.getData("key", i);
+                if(key.equals("FILENAME")){
+                    String tmpName = (String) dg.getData("value", i);
+                    spectraName += tmpName.substring(0,tmpName.lastIndexOf("."));
+                    break;
+                }
+            }
+            FileAnalysisReport tempRep = VoTableReader.analyze(spectra, FileAnalysisReport.ReportType.Details);
+            FileAnalysisReport.Part tempPart = tempRep.getPart(0);
+            FileAnalysisReport.Part addPart = tempPart.copy();
+            addPart.setDesc(spectraName.trim());
+            addPart.setFileLocationIndex(0);
+            addPart.setConvertedFileName(ServerContext.replaceWithPrefix(spectra));
+            addPart.setConvertedFileFormat(tempRep.getFormat());
+            addPart.setIndex(1);
+            addPart.setUiRender(FileAnalysisReport.UIRender.Chart);
+            addPart.setUiEntry(FileAnalysisReport.UIEntry.UseSpecified);
+            addPart.setChartTableDefOption(FileAnalysisReport.ChartTableDefOption.showChart);
+            addPart.setDefaultPart(true);
+
+
+
+            //TODO add error bars?
+            FileAnalysisReport.ChartParams cp = new FileAnalysisReport.ChartParams();
+            String xCol = VOSpectraModel.SPECTRA_FIELDS.WAVELENGTH.getKey();
+            String yCol = VOSpectraModel.SPECTRA_FIELDS.FLUX.getKey();
+            cp.setxAxisColName(xCol);
+            cp.setyAxisColName(yCol);
+            cp.setMode("markers");
+            addPart.setChartParams(cp);
+            retRep.addPart(addPart);
```

Comment:
We might want to change `FileAnalysisReport` so you can add a part to the beginning. You could do it now by getting the parts list, putting it at the beginning, cloning to an empty report, then setting each part in the right order. But that is too much work.